File: src/server/modules/ServerComponents/Frostbite.lua
local CollectionService = game:GetService("CollectionService")
local Players = game:GetService("Players")
local ReplicatedStorage = game:GetService("ReplicatedStorage")
local RunService = game:GetService("RunService")
local ServerStorage = game:GetService("ServerStorage")

local Knit = require(ReplicatedStorage.Packages.Knit)
local Signal = require(ReplicatedStorage.Packages.Signal)
local Trove = require(ReplicatedStorage.Packages.Trove)
local Component = require(ReplicatedStorage.Packages.Component)
local Logger = require(ReplicatedStorage.Source.Modules.Extensions.Logger)

local HealthService

local Frostbite = Component.new {
	Tag = "Frostbite";

	Extensions = {
		Logger,
	};
}

function Frostbite:Construct()
	self._trove = Trove.new()
end

function Frostbite:HeartbeatUpdate(deltaTime: number)
	local character: Model = self.Instance.Parent
	if not character then return end
	local player: Player = Players:GetPlayerFromCharacter(character)
	if not player then return end
	local data = player.Data
	if not data then return end
	if data.Temperature.Value >= 5 then return end

	local damage = math.min(self.Instance.Health, (self.Instance.MaxHealth/5)*deltaTime)
	self.Instance:TakeDamage(damage)

	if self.Instance.Health <= 0 then
		warn("blud froze to death innit")
		CollectionService:RemoveTag(self.Instance, "Frostbite")

		for _,child: Instance in self.Instance.Parent:GetChildren() do
			if not child:IsA("BasePart") then continue end
			if child.Transparency > 0.5 then continue end
			if child.Name == "Handle" then continue end

			task.spawn(function()
				child.Transparency = 1
				child.Anchored = true

				local frozen = child:Clone()
				frozen.Anchored = true
				frozen.Parent = workspace
				frozen:ClearAllChildren()
				frozen.CFrame = child.CFrame
				frozen.Size = child.Size
				frozen.Color = Color3.fromRGB(152, 194, 219)
				frozen.Material = "Ice"
				frozen.Transparency = 0.5

				local mesh = child:FindFirstChild("Mesh")
				if mesh then mesh:Clone().Parent = frozen end
			end)
		end

		if not self.Instance:IsDescendantOf(workspace.Dead) then
			HealthService.Kill(player)
		end
	end
end

function Frostbite:Start()
	HealthService = Knit.GetService("HealthService")
end

function Frostbite:Stop()
	self._trove:Destroy()
end

return Frostbite

File: src/server/components/injury/frostbite.ts
import { Component } from "@flamework/components";
import { OnStart, OnTick } from "@flamework/core";
import { Logger } from "@rbxts/log";
import { Workspace } from "@rbxts/services";
import { IdentityService } from "server/services/identity-service";
import { store } from "server/store";
import {
	selectResources,
	selectTemperature,
} from "shared/store/slices/players/slices/resources/selectors";
import { BaseInjury } from ".";
import { Character } from "../character/character";

const UPPER_TEMPERATURE_THRESHOLD = 5;
const DEATH_MESSAGE_TEMPLATE = "{Character} froze to death innit";
const ICE_COLOR = Color3.fromRGB(152, 194, 219);

@Component({
	tag: "Frostbite",
})
export class Frostbite extends BaseInjury implements OnStart, OnTick {
	public readonly name = "Frostbite";

	public constructor(
		character: Character,
		private logger: Logger,
		private identityService: IdentityService,
	) {
		super(character);
	}

	public onStart(): void {
		this.inflict();

		const player = this.character.getPlayer();
		const data = store.getState(selectResources(player.UserId));
		if (!data) error("no data");
		if (data.temperature === 0) {
			store.setTemperature(player.UserId, 15);
		}

		let skinColor = this.character.instance.GetAttribute("SkinColor");
		if (skinColor === undefined) {
			this.character.instance
				.GetAttributeChangedSignal("SkinColor")
				.Wait();
			skinColor = this.character.instance.GetAttribute("SkinColor")!;
		}
		const [h, s, v] = (skinColor as Color3).ToHSV();
		this.identityService.setCharacterSkinColor(
			this.character.instance,
			Color3.fromHSV(h, s / 2, v),
		);
	}

	public onTick(dt: number): void {
		if (!this.character.attributes.isAlive) return;
		const characterTemperature = store.getState(
			selectTemperature(this.character.getPlayer().UserId),
		);
		if (characterTemperature === undefined) return;
		if (characterTemperature > UPPER_TEMPERATURE_THRESHOLD) return;

		const humanoid = this.character.getHumanoid();
		humanoid.TakeDamage(this.calculateTickDamage(humanoid, dt));
		if (humanoid.Health > 0) return;

		this.character.instance.GetChildren().forEach((value) => {
			if (!value.IsA("BasePart")) return;
			if (value.Transparency > 0.5) return;
			if (value.Name === "Handle") return;
			this.freezePart(value);
		});

		this.character.kill();
		this.logger.Info(DEATH_MESSAGE_TEMPLATE, this.character.instance.Name);
	}

	private calculateTickDamage(humanoid: Humanoid, deltaTime: number): number {
		return math.min(humanoid.Health, (humanoid.MaxHealth * deltaTime) / 5);
	}

	private freezePart(part: BasePart): void {
		part.Transparency = 1;
		part.Anchored = true;

		const frozen = part.Clone();
		frozen.ClearAllChildren();
		frozen.Parent = Workspace;
		frozen.CFrame = part.CFrame;
		frozen.Size = part.Size;
		frozen.Color = ICE_COLOR;
		frozen.Material = Enum.Material.Ice;
		frozen.Transparency = 0.5;

		const mesh = part.FindFirstChild("Mesh");
		if (mesh) mesh.Clone().Parent = frozen;
	}
}
